### PR TITLE
Adding Amazon Glue Data Quality Service

### PR DESCRIPTION
**Pull request by gopidesupavan**
Adding Amazon Glue Data Quality Service. Doc, Hook, Operator, Sensor, Trigger, Waiter, Unit Test, System Test.

GlueDataQualityOperator: Create ruleset or update ruleset

GlueDataQualityRuleSetEvaluationRunOperator: Execute evaluations on multiple rulesets.

Sample Dag for creating ruleset and execute evaluation:

```
from datetime import datetime
from airflow import DAG
from airflow.providers.amazon.aws.operators.glue import GlueDataQualityOperator, \
    GlueDataQualityRuleSetEvaluationRunOperator

with DAG(
    dag_id="example_glue_data_quality",
    schedule="@once",
    start_date=datetime(2021, 1, 1),
    tags=["glue data quality ruleset evaluation"],
    catchup=False,
) as dag:

    rule_set_name = "test_rule_set"

    create_rule_set = GlueDataQualityOperator(
        task_id="create_rule_set",
        name=rule_set_name,
        ruleset='Rules = [ColumnLength "name" between 3 and 14]',
        data_quality_ruleset_kwargs={
            "TargetTable": {
                "TableName": "test_table",
                "DatabaseName": "test_default",
            }
        }
    )

    start_evaluation_run = GlueDataQualityRuleSetEvaluationRunOperator(
        task_id="start_evaluation_run",
        datasource={
            "GlueTable": {
                 "TableName": "test_table",
                "DatabaseName": "test_default",
            }
        },
        role="arn:aws:iam::{ACCOUNT_ID}:role/GlueDataQuality",
        rule_set_names=[rule_set_name]
    )
```

image

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 1]
```diff
@@ -430,3 +431,105 @@ def create_or_update_glue_job(self) -> str | None:
...
+            raise AirflowException(f"AWS Glue data quality ruleset {config['Name']} not exists to update")
+
+    def create_glue_data_quality_ruleset(self, config: dict[str, Any]) -> None:
+        if not self.has_data_quality_ruleset(config["Name"]):
+            self.log.info("Creating AWS Glue data quality ruleset with: %s", config)
+            self.conn.create_data_quality_ruleset(**config)
+        else:
+            raise AirflowException(
+                f"AWS Glue data quality ruleset {config['Name']} already exists with same name."
+            )
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/hooks/glue.py]: These function are only wrapper around boto3 API, I dont think they are necessary. Plus `update_data_quality_ruleset` and `create_data_quality_ruleset` have specific exception when the ruleset does not exist (for `update_data_quality_ruleset`) and when it already exist (for `create_data_quality_ruleset `). Thus, calling `has_data_quality_ruleset` is not needed

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 2]
```diff
@@ -430,3 +431,105 @@ def create_or_update_glue_job(self) -> str | None:
...
+
+    def display_result(self, result: dict[str, Any]) -> None:
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/hooks/glue.py]: Do we want it public? It think it is more a private function no?

Comment (vincbeck) [reply airflow/providers/amazon/aws/hooks/glue.py]: Some docstring would help as well to understand the function

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 3]
```diff
@@ -430,3 +431,105 @@ def create_or_update_glue_job(self) -> str | None:
...
+                results["ResultsNotFound"],
+            )
+
+        for result in results["Results"]:
+            rule_results = result["RuleResults"]
+            total_failed_rules = total_failed_rules + sum(
+                1
+                for result in rule_results
+                if result.get("Result") == "FAIL" or result.get("Result") == "ERROR"
+            )
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/hooks/glue.py]:
```suggestion
            total_failed_rules += len(filter(lambda result: result.get("Result") == "FAIL" or result.get("Result") == "ERROR", rule_results))
```

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 2 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/hooks/glue.py]: `log_results` would makes more sense as name?

File: airflow/providers/amazon/aws/operators/glue.py [hunk 4]
```diff
@@ -239,3 +243,253 @@ def on_kill(self):
...
+        maintained on each worker node).
+    :param region_name: AWS region_name. If not specified then the default boto3 behaviour is used.
+    :param verify: Whether or not to verify SSL certificates. See:
+        https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html
+    :param botocore_config: Configuration dictionary (key-values) for botocore client. See:
+        https://botocore.amazonaws.com/v1/documentation/api/latest/reference/config.html
+    """
+
+    aws_hook_class = GlueDataQualityHook
+    template_fields: Sequence[str] = ("name", "ruleset", "data_quality_ruleset_kwargs")
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/glue.py]:
```suggestion
    template_fields: Sequence[str] = ("name", "ruleset", "description", "data_quality_ruleset_kwargs")
```

File: airflow/providers/amazon/aws/operators/glue.py [hunk 5]
```diff
@@ -239,3 +243,253 @@ def on_kill(self):
...
+            self.hook.update_glue_data_quality_ruleset(config)
+            self.log.info("AWS Glue data quality ruleset updated successfully")
+        else:
+            self.hook.create_glue_data_quality_ruleset(config)
+            self.log.info("AWS Glue data quality ruleset created successfully")
+
+
+class GlueDataQualityRuleSetEvaluationRunOperator(AwsBaseOperator[GlueDataQualityHook]):
+    """
+    Once you have a ruleset definition (either recommended or your own), you call this operation to evaluate the ruleset against a data source (Glue table).
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/glue.py]: It feels weird to describe an operator with a sentence starting by "Once you have a". Suggestion: "Evaluate a ruleset against a data source (Glue table)".

File: airflow/providers/amazon/aws/operators/glue.py [hunk 6]
```diff
@@ -239,3 +243,253 @@ def on_kill(self):
...
+        glue_table = self.datasource.get("GlueTable", {})
+
+        if not glue_table.get("DatabaseName") or not glue_table.get("TableName"):
+            raise AttributeError("DataSource glue table must have DatabaseName and TableName")
+
+        not_found_ruleset = []
+
+        for ruleset_name in self.rule_set_names:
+            if not self.hook.has_data_quality_ruleset(ruleset_name):
+                not_found_ruleset.append(ruleset_name)
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/glue.py]:
```suggestion
        not_found_ruleset = [ruleset_name for ruleset_name in self.rule_set_names if not self.hook.has_data_quality_ruleset(ruleset_name)]
```

**Review by vincbeck — COMMENTED** [review]
Some comments but it is solid overall!

File: airflow/providers/amazon/aws/operators/glue.py (hunk 6 above)
Comment (gopidesupavan) [reply airflow/providers/amazon/aws/operators/glue.py]: Thank you @vincbeck , Appreciate your feedback 😄 . I will update changes.

File: airflow/providers/amazon/aws/operators/glue.py (hunk 5 above)
Comment (gopidesupavan) [reply airflow/providers/amazon/aws/operators/glue.py]: Agree, updated the changes.

File: airflow/providers/amazon/aws/operators/glue.py (hunk 4 above)
Comment (gopidesupavan) [reply airflow/providers/amazon/aws/operators/glue.py]: done changes.

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 2 above)
Comment (gopidesupavan) [reply airflow/providers/amazon/aws/hooks/glue.py]: make sense updated.

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 1 above)
Comment (gopidesupavan) [reply airflow/providers/amazon/aws/hooks/glue.py]: Agree , these are kind of wrappers, i have now removed those and calling directly inside operator execute method. `self.hook.conn.create_data_quality_ruleset(**config)`
`self.hook.conn.update_data_quality_ruleset(**config)`

Thank you.

File: airflow/providers/amazon/aws/operators/glue.py (hunk 6 above)
Comment (gopidesupavan) [reply airflow/providers/amazon/aws/operators/glue.py]: updated the changes.

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 2 above)
Comment (gopidesupavan) [reply airflow/providers/amazon/aws/hooks/glue.py]: Updated docstring for log_results function.

**Review by vincbeck — APPROVED** [review]

**Comment by gopidesupavan** [comment]
@vincbeck Thank you for all the suggestions, made all the changes please review. Is the docstring sufficient for the log_results?. please let me know happy to refine. 😄 .

**Comment by vincbeck** [comment]
> @vincbeck Thank you for all the suggestions, made all the changes please review. Is the docstring sufficient for the log_results?. please let me know happy to refine. 😄 .

Yep, it looks good, it definitely helps to understand the function!

**Pull request merged**